src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java

```diff
@@ -92,63 +113,61 @@ protected void init(MAPF_Instance instance, RunParameters parameters) {
         initPriority(instance);
 
         // distance between every vertex in the graph to each agent's goal
-        this.heuristic = new DistanceTableAStarHeuristic(instance.agents, instance.map);
+        if (parameters.aStarGAndH instanceof DistanceTableAStarHeuristic) {
+            this.heuristic = (DistanceTableAStarHeuristic) parameters.aStarGAndH;
+        }
+        else {
+            this.heuristic = new DistanceTableAStarHeuristic(instance.agents, instance.map);
+        }
     }
 
 
     @Override
     protected Solution runAlgorithm(MAPF_Instance instance, RunParameters parameters) {
-
+        boolean loopDetected = false;
         // each iteration of the while represents timestamp
         while (!(finished())) {
 
+
+            // loop detection
+            ArrayList<I_Location> currentConfigurations = new ArrayList<>();
+            for (Agent agent : this.currentLocations.keySet()) {
+                currentConfigurations.add(this.currentLocations.get(agent));
+            }
+            if (this.configurations.contains(currentConfigurations)) {
+                loopDetected = true;
+            }
+            else {
+                this.configurations.add(currentConfigurations);
+            }
+
+            if (loopDetected && DEBUG >= 2) {
+                System.out.println("LOOP DETECTED");
+                return null;
+            }
             if (checkTimeout()) {
                 return null;
             }
-
             this.timeStamp++;
 
-            updatePriorities(instance);
             // init agents that have not reached their goal
             this.unhandledAgents = new HashSet<>();
             for (Map.Entry<Agent, Double> entry : this.priorities.entrySet()) {
                 Agent agent = entry.getKey();
-                Double priority = entry.getValue();
-                if (priority != -1.0) {
-                    // the agent did not reach his goal
-                    this.unhandledAgents.add(agent);
-                }
+                this.unhandledAgents.add(agent);
```

Comment:
Are you sure this is OK?

Reply:
its weird I changed it long time ago and i don't see those lines of code (not at my project and not in the git)

Reply:
Must be from the mess I made with reverting stuff 😅